### PR TITLE
fix threads per block in inner persistent heuristics

### PR DESCRIPTION
**Pull request by liqiangxl**
1. In the test of softmax found an error with hidden size of 22024. It leads to threads per block larger than 1024 due to `ceilDiv`. In this PR, we update bdimx first then update batches per block.
2. Extended `FusionMagicSchedulerSoftmax_CUDA` to allow more test cases if `const bool test_all = true;` by default it is set to `false`.

### CODE REVIEW

**Comment by liqiangxl** [comment]
Perf changed only on one case with a feature size of 5140. Looks like all other cases are dividable.
Main branch (old), batch = 5, effective threads = 5140 / 4 / 5 = 257, launched = 288
```
ptxas info    : 307 bytes gmem
ptxas info    : Compiling entry function '_ZN11CudaCodeGen7kernel5ENS_6TensorINS_6__halfELi1ELi1EEENS0_IS1_Li2ELi2EEES2_S2_S2_S3_' for 'sm_80'
ptxas info    : Function properties for _ZN11CudaCodeGen7kernel5ENS_6TensorINS_6__halfELi1ELi1EEENS0_IS1_Li2ELi2EEES2_S2_S2_S3_
ptxas         .     0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 63 registers, 16 bytes smem, 464 bytes cmem[0], 8 bytes cmem[2]

blocks_per_sm= 3, warps_per_sm= 27, occupancy= 42.19%
NvFuserScheduler_LayerNormFusedOp_fp16___GRAPH/NvFuserScheduler_LayerNormFusedOp_fp16/8192/5140/manual_time         241 us          316 us           58 bytes_per_second=651.193G/s Red On Fastest Dim // Persistent Kernel //  // Iteration Domain:  // Inner Reduction Domain: cross block reduction / pad to warp / persistent batch - 5 / vectorize / factor 4/Launch_Parameters[block(1/1/288)/grid(1/1/8192)/1152]
```
With this change (new): batch = 6, effective threads = 5140 / 4 / 6 = 215, launched = 224
```
ptxas info    : 307 bytes gmem
ptxas info    : Compiling entry function '_ZN11CudaCodeGen7kernel5ENS_6TensorINS_6__halfELi1ELi1EEENS0_IS1_Li2ELi2EEES2_S2_S2_S3_' for 'sm_80'
ptxas info    : Function properties for _ZN11CudaCodeGen7kernel5ENS_6TensorINS_6__halfELi1ELi1EEENS0_IS1_Li2ELi2EEES2_S2_S2_S3_
ptxas         .     0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 71 registers, 16 bytes smem, 464 bytes cmem[0], 8 bytes cmem[2]
blocks_per_sm= 4, warps_per_sm= 28, occupancy= 43.75%
NvFuserScheduler_LayerNormFusedOp_fp16___GRAPH/NvFuserScheduler_LayerNormFusedOp_fp16/8192/5140/manual_time         210 us          284 us           67 bytes_per_second=746.823G/s Red On Fastest Dim // Persistent Kernel //  // Iteration Domain:  // Inner Reduction Domain: cross block reduction / pad to warp / persistent batch - 6 / vectorize / factor 4/Launch_Parameters[block(1/1/224)/grid(1/1/8192)/896]
```

image

File: csrc/scheduler/normalization.cpp [hunk 1]
```diff
@@ -598,10 +598,9 @@ std::shared_ptr<ReductionParams> innerPersistentHeuristic(
     // than five percent.
     if (batches_per_block_inner_reduction >
         batches_per_block_inner_reduction_max) {
-      batches_per_block_inner_reduction /= 2;
-      bdimx = ceilDiv(
-          inner_most_dimension_numel,
-          inner_reduction_unroll_factor * batches_per_block_inner_reduction);
+      bdimx *= 2;
```
[the remark this answers is not in the document]
Comment (liqiangxl) [reply csrc/scheduler/normalization.cpp:596]: updated

File: test/test_gpu1.cpp [hunk 2]
```diff
@@ -7465,43 +7465,68 @@ TEST_F(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
...
+        cg_outputs,
+        {aten_input},
+        {aten_output},
+        __LINE__,
+        __FILE__,
+        "",
+        lparams);
+  };
+
+  const bool test_all = false;
```
[the remark this answers is not in the document]
Comment (liqiangxl) [reply test/test_gpu1.cpp]: yes. I tested 2 x 4096 cases. Changed to a smaller set but includes different types of cases.

**Comment by naoyam** [comment]
>Looks like all other cases are dividable.

Are you sure? Can you please confirm it?

**Comment by liqiangxl** [comment]
> 



> Are you sure? Can you please confirm it?

You are right. There are also other cases but there is no noticable change in perf.
image

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the fix!

**Pull request merged**